src/SPH.cpp

```diff
@@ -632,21 +637,16 @@ void update_scalar_function(Float3* pos, Neighbor_Data* neighbor_data, float* sc
 			dy = pos->y[particle] - pos->y[neighbor_ind];
 			dz = pos->z[particle] - pos->z[neighbor_ind];
 
-			dist = std::sqrt(dx*dx + dy*dy + dz*dz);
+			dist = sqrt(dx*dx + dy*dy + dz*dz);
 			q = dist / D_RAD;
-			
-			//this is an approximation 
-			dist = dist == 0.f ? 0.000000000000000000000000000001f : dist;
-
-			//Compute the derivitive of the kernel function
-			if (q >= 0 || q <= 0.5f)
-				kernel_derive = (-12.f*q + 18.f*q*q);
-			else if (q > 0.5f || q <= 1.0f)
-				kernel_derive = -6.0f*(1 - q)*(1 - q);
-			else
-				kernel_derive = 0.0f;
-
-			scalar_values[particle*D_MAX_NR_OF_NEIGHBORS + neighbor] = kernel_derive/(dist*radii);
+			q_2 = q*q;
+
+			// length is always equal or smaller to D_RAD => implicit intervall between [0, 1]
+			kernel_derive = (1.0f / (search_area*pi))*(-3.0f*q + 2.25f*q_2);
+
+			scalar_value = kernel_derive*(1.0f /  (search_area*dist));
```

Comment:
remove the 1.0f from line 645 to :

kernel_derive /  (search_area*dist));

instead?